### PR TITLE
Conversion from DiningStack external eateries to GraphQL

### PR DESCRIPTION
**Pull request by ilikesymmetry**
Added a missing key handler for keys that statics eateries may not have.  Created mechanism for taking external eatery hours and convert to our current graphql style.  Added try-excepts to handle missing information in external eateries. #18

### CODE REVIEW

File: data.py [hunk 1]
```diff
@@ -1,31 +1,25 @@
-import constants
-from datetime import datetime
+from constants import *
```
[the remark this answers is not in the document]
Comment (young-k) [reply data.py]: @cswenberg hound is correct, try and avoid `*` imports and just name all the imports instead

File: data.py [hunk 2]
```diff
@@ -137,33 +131,128 @@ def parse_food_items(item_list, eatery_id):
...
+    for i in range(100):
+      test_id = 100 - i
+      if test_id not in eateries:
+        return test_id
+
+def parse_eatery_type(eatery):
+  try:
+    return eatery['eateryTypes'][0]['descr']
+  except Exception:
+    return 'unknown'
```
Comment (young-k) [inline data.py]: better exception message, we're gonna add logging to these in the future

File: data.py [hunk 3]
```diff
@@ -137,33 +131,128 @@ def parse_food_items(item_list, eatery_id):
...
+      print('weekday key: ' + hours['weekday'] + ' not in constants WEEKDAYS!')
+      continue
+    for weekday in WEEKDAYS[hours['weekday']]:
+      if weekday not in weekdays:
+        weekdays[weekday] = hours['events']
+
+  global operating_hours
+  global today
+  new_operating_hours = []
+  for i in range(8):
```
Comment (young-k) [inline data.py]: why 8, and can you define this in a constant at the top?

File: data.py [hunk 4]
```diff
@@ -137,33 +131,128 @@ def parse_food_items(item_list, eatery_id):
...
+    new_date = date(today.year, today.month, today.day + i)
+    weekday = new_date.weekday()
+    new_events = weekdays[weekday] if weekday in weekdays.keys() else []
```
Comment (young-k) [inline data.py]: this should be: `new_events = weekdays.get(weekday, [])`

File: data.py [hunk 5]
```diff
@@ -137,33 +131,128 @@ def parse_food_items(item_list, eatery_id):
...
+def safe_parse(obj, key, default=''):
+  try:
+    value = obj[key]
+    if value is None:
+      return default
+    return value
+  except Exception:
+    return default
+
+def resolve_id(obj):
```
Comment (young-k) [inline data.py]: can we add comments to all of these functions explaining what they do, ex:

```
def resolve_id(obj):
  """Returns [SOMETHING]
  [OPTIONAL EXPLANATION IF NECESSARY]

  Args:
    obj: [INSERT HERE]
  """
```

Refer to https://github.com/google/styleguide/blob/gh-pages/pyguide.md#383-functions-and-methods

**Review by young-k — CHANGES_REQUESTED** [review]
Pretty bulky changes, try and get these in by Wednesday!

File: data.py (hunk 2 above)
Comment (ilikesymmetry) [reply data.py]: iOS uses an enum from the value for 'eateryType' where unknown is a valid value when we can't determine another value otherwise

**Comment by ilikesymmetry** [comment]
@young-k changes made

File: data.py [hunk 6]
```diff
@@ -1,31 +1,26 @@
-import constants
-from datetime import datetime
+from constants import ASSET_BASE_URL, CORNELL_DINING_URL, NUM_DAYS_STORED, STATIC_EATERIES_URL, \
+    PAY_METHODS, UPDATE_DELAY, WEEKDAYS
```
Comment (kq-li) [inline data.py]: for long imports, prefer
```
from constants import (
    ASSET_BASE_URL,
    CORNELL_DINING_URL,
    ...
)
```

or just `from constants import *`

File: data.py [hunk 7]
```diff
@@ -62,7 +57,7 @@ def parse(data_json):
     eateries[new_eatery.id] = new_eatery
 
 def get_image_url(slug):
-  return constants.ASSET_BASE_URL + 'eatery-images/' + slug + '.jpg'
+  return ASSET_BASE_URL + 'eatery-images/' + slug + '.jpg'
```
Comment (kq-li) [inline data.py]: refactor this, `IMAGES_URL` perhaps?

Comment (kq-li) [reply data.py]: alternatively, consider a utils class with a function to generate an image URL from a given name

File: data.py [hunk 8]
```diff
@@ -137,33 +132,133 @@ def parse_food_items(item_list, eatery_id):
...
+        name=safe_parse(eatery, 'name'),
+        name_short=safe_parse(eatery, 'nameshort'),
+        operating_hours=parse_static_op_hours(eatery['operatingHours'], new_id),
+        payment_methods=parse_payment_methods(eatery['payMethods']),
+        phone=parse_phone(eatery['contactPhone'], eatery['name']),
+        slug=safe_parse(eatery, 'slug')
+    )
+    eateries[new_eatery.id] = new_eatery
+
+def safe_parse(obj, key, default=''):
```
Comment (kq-li) [inline data.py]: A++ intuition, the developers of python agreed with you that this is useful and bestowed us with `dict.get(key, default)` which is pretty much the best thing ever invented

File: data.py [hunk 9]
```diff
@@ -137,33 +132,133 @@ def parse_food_items(item_list, eatery_id):
...
+    return value
+  except Exception:
+    return default
+
+def resolve_id(eatery):
+  """Returns a new id (int) for an external eatery
+  If the eatery does not have a provided id, we need to make one for it.
+  Works by going backwards from 99 and making sure the id is not already takenself.
+  Should feasibly never conflict with cornell's eateries (numbered ascending, current max is < 30).
+  """
```
Comment (kq-li) [inline data.py:211]: prefer triple single quotes `'''` over doubles `"""`

File: data.py [hunk 10]
```diff
@@ -137,33 +132,133 @@ def parse_food_items(item_list, eatery_id):
...
+  try:
+    id = eatery['id']
+    return id
+
+  except Exception:
+    global eateries
+    for i in range(100):
+      test_id = 100 - i
+      if test_id not in eateries:
+        return test_id
```
Comment (kq-li) [inline data.py]: is this for debugging? what's the significance of 100?

File: data.py [hunk 11]
```diff
@@ -137,33 +131,128 @@ def parse_food_items(item_list, eatery_id):
...
+def parse_eatery_type(eatery):
+  try:
+    return eatery['eateryTypes'][0]['descr']
+  except Exception:
+    return 'unknown'
+
+def parse_static_op_hours(hours_list, eatery_id):
+  weekdays = {}
+  for hours in hours_list:
+    if hours['weekday'] not in WEEKDAYS.keys():
```
[the remark this answers is not in the document]
Comment (kq-li) [reply data.py]: `x not in dict.keys()` and `x not in dict` are equivalent in python

**Review by kq-li — COMMENTED** [review]
Good stuff -- just some minor fixes needed, but otherwise lgtm

File: constants.py [hunk 12]
```diff
@@ -1,3 +1,25 @@
...
+WEEKDAYS = {
+    'monday-wednesday': [0, 1, 2],
+    'monday-thursday': [0, 1, 2, 3],
+    'monday-friday': [0, 1, 2, 3, 4],
+    'thursday': [3],
+    'friday': [4],
+    'saturday': [5],
+    'sunday': [6]
+}
+NUM_DAYS_STORED = 8
```
Comment (alanna-zhou) [inline constants.py]: why is this 8?

Comment (ilikesymmetry) [reply constants.py]: Cornell Dining's json stores today and the next week of eatery information, so our graphql db stores 8 days worth of information at any point in time

File: data.py (hunk 9 above)
Comment (young-k) [reply data.py:211]: @kq-li this isn't correct, docstrings are in double quotes

Comment (young-k) [reply data.py:211]: @cswenberg can you redo this change?

File: data.py [hunk 13]
```diff
@@ -129,41 +128,125 @@ def parse_food_items(item_list, eatery_id):
...
+  try:
+    return eatery['eateryTypes'][0]['descr']
+  except Exception:
+    return 'Unknown'
+
+def parse_static_op_hours(hours_list, eatery_id):
+  weekdays = {}
+  for hours in hours_list:
+    if hours['weekday'] not in WEEKDAYS:
+      print('weekday key: ' + hours['weekday'] + ' not in constants WEEKDAYS!')
```
Comment (young-k) [inline data.py]: use a format string here

File: data.py [hunk 14]
```diff
@@ -129,41 +128,125 @@ def parse_food_items(item_list, eatery_id):
...
+        phone=eatery.get('contactPhone', 'N/A'),
+        slug=eatery.get('slug', '')
+    )
+    eateries[new_eatery.id] = new_eatery
+
+def resolve_id(eatery):
+  '''Returns a new id (int) for an external eatery
+  If the eatery does not have a provided id, we need to make one for it.
+  Simply gets the maximum id currently in our eateries and makes new id +1
+  '''
```
Comment (young-k) [inline data.py]: change this back to double quotes

File: constants.py (hunk 12 above)
Comment (young-k) [reply constants.py]: make this more verbose, `NUM_DAYS_STORED_IN_DB`

Comment (young-k) [reply constants.py]: alphabetize this file as well

**Review by young-k — CHANGES_REQUESTED** [review]
Almost there, minor changes this time, will approve afterwards.

File: data.py (hunk 6 above)
Comment (young-k) [reply data.py]: @kq-li we're trying to avoid `*` imports, it's not good practice in general

**Comment by young-k** [comment]
Another comment, please press the `Resolve` button on comments after you resolve them!

File: data.py (hunk 9 above)
Comment (kq-li) [reply data.py:211]: whoops, you're right, my bad

**Review by young-k — APPROVED** [review]
Great job with this! *Try* rebasing and squash the previous commits to fix PR stuff, if it doesn't work just `rebase --abort` and merge in!

**Pull request merged**